tools/clang/lib/CodeGen/CGHLSLMS.cpp

```diff
@@ -6060,16 +6060,37 @@ void CGMSHLSLRuntime::EmitHLSLOutParamConversionInit(
     const ParmVarDecl *Param = FD->getParamDecl(i);
     const Expr *Arg = E->getArg(i+ArgsToSkip);
     QualType ParamTy = Param->getType().getNonReferenceType();
-
+    bool RValOnRef = false;
     if (!Param->isModifierOut()) {
-      if (!ParamTy->isAggregateType() || hlsl::IsHLSLMatType(ParamTy))
-        continue;
+      if (!ParamTy->isAggregateType() || hlsl::IsHLSLMatType(ParamTy)) {
+        if (Arg->isRValue() && Param->getType()->isReferenceType()) {
+          // RValue on a reference type.
```

Comment:
Let's add a TODO comment here about evolving this to warn then fail in future language versions.